### ISSUE TITLE
Redirects are performed sequential even when using the parallel sendAll(), resulting in a bad performance

### ISSUE THREAD

**Issue by paslandau** [guzzle/guzzle#804]
I'm switching from my own Curl-Client implementation to Guzzle and experienced a bad performance while running a custom login script with multiple accounts (using sendAll to send requests in parallel).

After some debugging I noticed, that Guzzle will handle redirects "manually" instead of leaving it to curl. So

```
$client->get('http://github.com', ['allow_redirects' => true]);
```

is not the same as 

```
$client->get('http://github.com', ['config' => ['curl' => [CURLOPT_FOLLOWLOCATION => true]]]);
```

That's actually not a problem, as long as you do not send parallel requests. This is what seems to happen if you do:

```
- Send n requests in parallel
- receive first response
- propagate event to the RedirectSubscriber
- check status code
- if it's a 3xx, clone the original request and **immediately perform a sequential(!) redirect**
```

Corresponding file: ..\guzzlehttp\guzzle\src\Subscriber\Redirect.php

I think this kinda defeats the purpose of sending requests in bulk :)

Using `CURLOPT_FOLLOWLOCATION` solves this problem as the response propagated to the RedirectSubscriber will be the final response (since curl internally handled all redirects already).

Benchmark-Script:

```
<?php
$url = "http://de.wikipedia.org/";
$urlNoRedirect = "http://de.wikipedia.org/wiki/Wikipedia:Hauptseite";
$con = 20;
for($i=0; $i<50; $i++){
    $requestsRedirectDefault[] = $client->createRequest("GET",$url, ['allow_redirects' => true]);
    $requestsRedirectCurl[] = $client->createRequest("GET",$url,['config' => ['curl' => [CURLOPT_FOLLOWLOCATION => true]]]);
}

$client = new Client();
$start = microtime(true);
$results = GuzzleHttp\batch($client,$requestsRedirectDefault,['parallel' => $con]);
$t = microtime(true) - $start;
echo "$t s: Requesting '$url' (redirect to '$urlNoRedirect') with guzzle default redirect\n";

$client = new Client();
$start = microtime(true);
$results = GuzzleHttp\batch($client,$requestsRedirectCurl,['parallel' => $con]);
$t = microtime(true) - $start;
echo "$t s: Requesting '$url' (redirect to '$urlNoRedirect')) with CURLOPT_FOLLOWLOCATION redirect\n";
?>
```

Result:

```
2.3671360015869 s: Requesting 'http://de.wikipedia.org/' (redirect to 'http://de.wikipedia.org/wiki/Wikipedia:Hauptseite') with guzzle default redirect
1.2240698337555 s: Requesting 'http://de.wikipedia.org/' (redirect to 'http://de.wikipedia.org/wiki/Wikipedia:Hauptseite')) with CURLOPT_FOLLOWLOCATION redirect
```

---

I actually wouldn't bother and just happily use `CURLOPT_FOLLOWLOCATION` - but then I encountered another issue when using cookies:

It seems that Guzzle does not use curl's `CURLOPT_COOKIEJAR` / `CURLOPT_COOKIEFILE` to handle cookies. Unfortunately, that seems to be required when using `CURLOPT_FOLLOWLOCATION` because my login script fails until i manually create a file for `CURLOPT_COOKIEJAR` / `CURLOPT_COOKIEFILE`.

Then again: This feels kinda wrong and it would be great to have Guzzle deal with that for me :) 

---

Any ideas what could be done about this?
